Polygonize with watershed

Gedeon-m-gedus opened the pull request:
## Add Watershed-based Polygonization for Improved Field Instance Separation

Adding hierarchical watershed-based instance segmentation to improve handling of touching/overlapping agricultural fields during polygonization.

#### Changes
- **New module**: `ftw_tools/postprocess/watershed_polygonize.py` with watershed algorithm implementation
- **New CLI command**: ftw inference `polygonize-watershed` with watershed-specific parameters (`--t_ext, --t_bound`).
 `t_ext` is the threshold for binarizing fields extent and `t_bound` is threshold for cutting the hierarchical watershed tree.
- **Fallback support**: Graceful handling when `fiboa_cli.parquet` is unavailable
- **Isolated implementation**: Separate from existing polygonization to minimize conflicts

#### Key Features
- Better separation of touching fields using hierarchical watershed
- Configurable thresholds for extent and boundary detection. While the expected input field extent should be a probability map, this should work fine with ftw outputs because the function binarizes the input (background vs fields_and_boundaries).
- Same output format compatibility as standard polygonization

#### Dependencies
- It needs [higra](https://pypi.org/project/higra/) package, I installed it manually by `conda install higra -c conda-forge`. I think pip should work also `pip install higra`
- All existing dependencies remain unchanged

#### Example Usage

```bash
ftw inference polygonize-watershed model_output_fields.tif -o output_fields_instances.geojson
```

Following images show sample visualization from the tests I did;

Input image:
<img width="1099" height="447" alt="Screenshot 2025-09-11 at 8 53 34 PM" src="https://github.com/user-attachments/assets/c016bd76-3156-44ee-b70f-5c4fd884c865" />

Predicted fields with 3classes model:
<img width="1099" height="447" alt="Screenshot 2025-09-11 at 8 53 43 PM" src="https://github.com/user-attachments/assets/ac1fcc6c-0f2a-4923-b23a-aa266fa5c077" />

Instance segmentation with this algorithm:
<img width="1099" height="447" alt="Screenshot 2025-09-11 at 8 53 50 PM" src="https://github.com/user-attachments/assets/c6378fec-215a-4681-8e65-fdbc5e1ff57c" />

#### Performance Note
⚠️ Slower than standard polygonization - The hierarchical watershed algorithm is computationally more intensive but provides good field separation for complex cases.

Commit:
create instances with watershed

hannah-rae commented:
I tested this with one of our prediction tifs in South America. Some usage feedback: 
- I agree with @cholmes that it should be unified with the `ftw inference polygonize` command
- There should be a progress bar like we have in the baseline polygonize (I ran it and was like... is it going?)
- It seems like there could be a max size of tif this will work with. I ran the baseline method on my tif and it took ~1 minute and found 80.5k polygons (for an area of ~350 km x 350 km). The watershed method ran for >15 minutes and then died with no error - I'm guessing the process ran out of memory or something. Not sure we can do anything about this 🤷‍♀️ 

I ran it on the austria example in the repo and that worked (took ~7 minutes compared to <1min for baseline). There are some minor differences, but I don't see any major improvement... it seems like some polygons might be better, some might be worse. 

<img width="1209" height="720" alt="image" src="https://github.com/user-attachments/assets/5e339073-262d-4efe-9cf7-58bf8d2d1834" />

baseline/GDAL polygonize:
<img width="1209" height="720" alt="image" src="https://github.com/user-attachments/assets/ddaf1dcb-a365-481f-8d7d-b65f9cc6c432" />

watershed polygonize:
<img width="1209" height="720" alt="image" src="https://github.com/user-attachments/assets/22ecd7ba-a1af-45b7-acc9-4d7760d451be" />

Gedeon-m-gedus committed:
combining polygonizations in one command

Gedeon-m-gedus committed:
Progress bar

Gedeon-m-gedus commented:
@cholmes @hannah-rae 
Thanks for the feedback; I have updated the cli to be single command 
- Simple: `ftw inference polygonize tests/data-files/mask.tif --algorithm simple -o test_simple.geojson`
- Watershed: `ftw inference polygonize tests/data-files/mask.tif --algorithm watershed -o test_watershed.geojson`

Simple is the default, so normal command should run (eg: `ftw inference polygonize tests/data-files/mask.tif -o test_simple.geojson`)

I also added progress bar (especially since watershed is slow)

RE: Max size, I don't have a number but we can find the limit experimentally (it could also depends on user hardware)

hannah-rae commented:
Thanks @Gedeon-m-gedus! I tested the updated code and it works for me. I see the main branch needs to be merged and there are some pytest issues, but I approve merging once those are fixed.

hannah-rae reviewed (changes requested):
Good for me once the pytests are passing

asmithml committed:
chg: update to new pixi.toml structure and add higra

hannah-rae commented:
@Gedeon-m-gedus Don't worry about the pytest issue as this is fixed from another PR now. I think you mainly need to resolve the conflicts now before merging.

cannox227 commented:
Hi @hannah-rae, what’s the status of the watershed polygonization? Is this method still under development?
